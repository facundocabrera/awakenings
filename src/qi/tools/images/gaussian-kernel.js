/**
 * @source https://fiveko.com/gaussian-filter-in-pure-javascript/
 */
export const kernel = (σ, dim) => {
  const sqrtSigmaPi2 = Math.sqrt(Math.PI * 2.0) * σ;
  const s2 = 2.0 * σ ** 2;
  const kernel = new Float32Array(dim - !(dim & 1)); // Make it odd number
  const half = parseInt(kernel.length / 2);
  let sum = 0.0;

  for (let j = 0, i = -half; j < kernel.length; i++, j++) {
    kernel[j] = Math.exp(-(i ** 2) / s2) / sqrtSigmaPi2;
    sum += kernel[j];
  }

  // Normalize the gaussian kernel to prevent image darkening/brightening
  for (let i = 0; i < dim; i++) {
    kernel[i] /= sum;
  }

  return kernel;
};

/**
 * 1. I could call the function using 2 different sigma values, for now I
 *    prefer simplicity.
 * 2. It always return a odd number of elements to represent the maximun
 *    value of the distribution (left and right values are mirrored).
 */
export const kernel2D = (σ, width, height) => {
  const x = kernel(σ, width);

  if (x.length != width)
    throw Error(
      "You are trying to create a matrix but kernel and dimensions values does not match."
    );

  const matrix = new Array(height);
  for (let h = 0; h < height; h++) {
    const row = new Array(width);

    for (let w = 0; w < width; w++) {
      row[w] = x[w] * x[h];
    }

    matrix[h] = row;
  }

  return matrix;
};

// @TODO candidate to be called replicate and use density as the replication value
export const densify = (matrix, [width, height, density]) => {
  const result = new Array(height);

  for (let h = 0; h < height; h++) {
    const row = new Array(width * density);

    for (let w = 0; w < width; w++) {
      for (let d = 0; d < density; d++) {
        row[w * density + d] = matrix[h][w];
      }
    }

    result[h] = row;
  }

  return result;
};
