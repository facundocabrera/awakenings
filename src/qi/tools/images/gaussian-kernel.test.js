/**
 * Notes:
 * 1. Building a 2D kernel from 2 equal evaluation make the result keep the
 *    normalization properties (sum eq 1)
 */
import { densify, kernel2D as k2d } from "./gaussian-kernel";

// import { sum_matrix } from "../geometry/vector";

describe("kernel2D", () => {
  test("sigma 0.54", () => {
    const K = k2d(0.54, 3, 3);

    expect(K).toMatchSnapshot();

    // console.log(sum_matrix(K)); ~1
  });

  test("sigma 0.66", () => {
    const K = k2d(0.66, 3, 3);

    expect(K).toMatchSnapshot();

    // console.log(sum_matrix(K)); ~1
  });

  test("sigma 0.33", () => {
    const K = k2d(0.33, 3, 3);

    expect(K).toMatchSnapshot();

    // console.log(sum_matrix(K)); ~1
  });

  test("sigma 0.618", () => {
    const K = k2d(0.618, 3, 3);

    expect(K).toMatchSnapshot();

    console.log(K);
  });
});

describe("densify", () => {
  test("sigma 0.54", () => {
    const K = k2d(0.54, 3, 3);
    const K2 = densify(K, [3, 3, 2]);

    expect(K2).toMatchSnapshot();
  });

  test.only("kernel 1", () => {
    const K = [
      [1, 0, -1],
      [0, 0, 0],
      [-1, 0, 1],
    ];

    const kd = densify(K, [3, 3, 4]);

    expect(kd).toMatchSnapshot();
  });
});
