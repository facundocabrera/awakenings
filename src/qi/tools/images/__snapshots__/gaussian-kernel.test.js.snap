// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`densify kernel 1 1`] = `
Array [
  Array [
    1,
    1,
    1,
    1,
    0,
    0,
    0,
    0,
    -1,
    -1,
    -1,
    -1,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    -1,
    -1,
    -1,
    -1,
    0,
    0,
    0,
    0,
    1,
    1,
    1,
    1,
  ],
]
`;

exports[`densify sigma 0.54 1`] = `
Array [
  Array [
    0.01752041518399694,
    0.01752041518399694,
    0.09732387878174542,
    0.09732387878174542,
    0.01752041518399694,
    0.01752041518399694,
  ],
  Array [
    0.09732387878174542,
    0.09732387878174542,
    0.5406228837416762,
    0.5406228837416762,
    0.09732387878174542,
    0.09732387878174542,
  ],
  Array [
    0.01752041518399694,
    0.01752041518399694,
    0.09732387878174542,
    0.09732387878174542,
    0.01752041518399694,
    0.01752041518399694,
  ],
]
`;

exports[`kernel2D sigma 0.33 1`] = `
Array [
  Array [
    0.00009875321442009977,
    0.009739958762047252,
    0.00009875321442009977,
  ],
  Array [
    0.009739958762047252,
    0.9606451520941306,
    0.009739958762047252,
  ],
  Array [
    0.00009875321442009977,
    0.009739958762047252,
    0.00009875321442009977,
  ],
]
`;

exports[`kernel2D sigma 0.54 1`] = `
Array [
  Array [
    0.01752041518399694,
    0.09732387878174542,
    0.01752041518399694,
  ],
  Array [
    0.09732387878174542,
    0.5406228837416762,
    0.09732387878174542,
  ],
  Array [
    0.01752041518399694,
    0.09732387878174542,
    0.01752041518399694,
  ],
]
`;

exports[`kernel2D sigma 0.66 1`] = `
Array [
  Array [
    0.037683565533427554,
    0.11875542180484899,
    0.037683565533427554,
  ],
  Array [
    0.11875542180484899,
    0.3742440506468938,
    0.11875542180484899,
  ],
  Array [
    0.037683565533427554,
    0.11875542180484899,
    0.037683565533427554,
  ],
]
`;

exports[`kernel2D sigma 0.618 1`] = `
Array [
  Array [
    0.030745624682337658,
    0.11385305911130494,
    0.030745624682337658,
  ],
  Array [
    0.11385305911130494,
    0.42160532443007526,
    0.11385305911130494,
  ],
  Array [
    0.030745624682337658,
    0.11385305911130494,
    0.030745624682337658,
  ],
]
`;
